File: src/jray/math/VectMatrix.java
package jray.math;

import jray.common.*;

public class VectMatrix {

    public static void multiply(Matrix4 m, Vect3 v, Vect3 erg) {
        double[][] md = m.getData();
        double[] vd = v.getData();
        double[] ed = erg.getData();

        double fourth = md[3][0] * vd[0] + md[3][1] * vd[1] + md[3][2] * vd[2] + md[3][3];

        if (fourth != 0) {
            double x = (md[0][0] * vd[0] + md[0][1] * vd[1] + md[0][2] * vd[2] + md[0][3]) / fourth;
            double y = (md[1][0] * vd[0] + md[1][1] * vd[1] + md[1][2] * vd[2] + md[1][3]) / fourth;
            double z = (md[2][0] * vd[0] + md[2][1] * vd[1] + md[2][2] * vd[2] + md[2][3]) / fourth;

            ed[0] = x;
            ed[1] = y;
            ed[2] = z;
        }
    }

    public static void multiply(Vect3 v, Matrix4 m, Vect3 erg) {
        double[] vd = v.getData();
        double[][] md = m.getData();
        double[] ed = erg.getData();

        double fourth = md[0][3] * vd[0] + md[1][3] * vd[1] + md[2][3] * vd[2] + md[3][3];

        if(fourth != 0) {
            double x = (md[0][0] * vd[0] + md[1][0] * vd[1] + md[2][0] * vd[2] + md[3][0]) / fourth;
            double y = (md[0][1] * vd[0] + md[1][1] * vd[1] + md[2][1] * vd[2] + md[3][1]) / fourth;
            double z = (md[0][2] * vd[0] + md[1][2] * vd[1] + md[2][2] * vd[2] + md[3][2]) / fourth;

            ed[0] = x;
            ed[1] = y;
            ed[2] = z;
        }
    }
}

File: JRayXLib/JRayXLib/Math/VectMatrix.cs
using JRayXLib.Shapes;

namespace JRayXLib.Math
{
    public class VectMatrix
    {
        public static void Multiply(Matrix4 m, Vect3 v, ref Vect3 erg)
        {
            double[,] md = m.GetData();
            double fourth = md[3, 0]*v.X + md[3, 1]*v.Y + md[3, 2]*v.Z + md[3, 3];

            if (System.Math.Abs(fourth - 0) > Constants.EPS)
            {
                double x = (md[0, 0]*v.X + md[0, 1]*v.Y + md[0, 2]*v.Z + md[0, 3])/fourth;
                double y = (md[1, 0]*v.X + md[1, 1]*v.Y + md[1, 2]*v.Z + md[1, 3])/fourth;
                double z = (md[2, 0]*v.X + md[2, 1]*v.Y + md[2, 2]*v.Z + md[2, 3])/fourth;

                erg.X = x;
                erg.Y = y;
                erg.Z = z;
            }
        }

        public static void Multiply(Vect3 v, Matrix4 m, ref Vect3 erg)
        {
            double[,] md = m.GetData();

            double fourth = md[0, 3]*v.X + md[1, 3]*v.Y + md[2, 3]*v.Z + md[3, 3];

            if (System.Math.Abs(fourth - 0) > Constants.EPS)
            {
                double x = (md[0, 0]*v.X + md[1, 0]*v.Y + md[2, 0]*v.Z + md[3, 0])/fourth;
                double y = (md[0, 1]*v.X + md[1, 1]*v.Y + md[2, 1]*v.Z + md[3, 1])/fourth;
                double z = (md[0, 2]*v.X + md[1, 2]*v.Y + md[2, 2]*v.Z + md[3, 2])/fourth;

                erg.X = x;
                erg.Y = y;
                erg.Z = z;
            }
        }
    }
}